### PR TITLE
Bulged Frustums in BinaryCompactObject Domain

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

Adds the `sphericity` parameter to Frustum as well as to the BinaryCompactObject Domain.
Addresses some of issue #3545 

### Upgrade instructions

The `EnvelopingCube` in the BinaryCompactObject Domain will now take a `Sphericity` parameter, which can be
set to any value between 0 and 1, inclusive. A value of 0 corresponds to an EnvelopingCube with a flat surface, and
a value of 1 corresponds to an EnvelopingCube with a spherical surface.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/CoordinateMaps/Frustum.hpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.hpp]: typo: a set of

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (nilsvu) [inline src/Domain/Creators/BinaryCompactObject.cpp:378]: Haven't tested yet, but I believe line 420 also needs `frustum_sphericity_`, no?

Comment (mar-celine) [reply src/Domain/Creators/BinaryCompactObject.cpp:378]: Yes, there definitely should be a matching `frustum_sphericity` for the layer immediately following this one! good catch :slightly_smiling_face:

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp:111]: Could you add a comment why this max() * sqrt(3) is the correct radius?

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: I suggest you set `physical_{xyz}` like you did before, and then add to it in a `if (sphericity > 0.)` clause. Then there's no overhead from computing the correction if it's multiplied by zero anyway.

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: I think this can be `const Frustum& map`

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: also can be `const std::array<double, 3>&`

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp:245]: This is pretty nice, using the non-bulged inverse as initial guess for the rootfind. Could you add a brief comment pointing this out? I think it's always useful to leave comments in the code about important decisions like this that can easily be overlooked.

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp:228]: I believe this struct could be used to rootfind the inverse for _any_ map. Not necessarily for this PR, but you can consider factoring it out so you can re-use it.

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp:346]: In cases where you need to allocate a lot of temporaries like here, you can consider using a `TempBuffer`. It reduces the number of memory allocations, and possibly also the amount of memory used. Up to you if you want to do this here (I don't know if performance in these maps is relevant for the evolution code, it certainly isn't in the elliptic code). I think it's a good habit to adopt in future code though. Here's an example:

```cpp
  TempBuffer<tmpl::list<
      ::Tags::TempScalar<0, ReturnType>,
      ::Tags::TempScalar<1, ReturnType>, ...>>  // You can also use higher-rank tensor tags
      buffer(get_size(xi));
  // Reference tensor (components) in the buffer, giving them names
  auto& one_over_mag_flat = get(get<::Tags::TempScalar<0, ReturnType>>(buffer));
  auto& r_over_mag_flat = get(get<::Tags::TempScalar<1, ReturnType>>(buffer));
  // ...
  // Now you can use the memory for computations
  one_over_mag_flat  = 1.0 / sqrt(square(flat_frustum_x) + square(flat_frustum_y) + square(flat_frustum_z));
  r_over_mag_flat = radius_ * one_over_mag_flat;
  // It's also useful to re-purpose components of buffer when you don't need them anymore.
  // If you need a component in the buffer only temporarily, use scopes:
  {
    auto& xi_column_prefactor = get(get<::Tags::TempScalar<0, ReturnType>>(buffer));
  } // `xi_column_prefactor` is only available in the scope. Outside, it's clear that the part of the buffer can be re-used.
```

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp:47]: Add an assert that `0. <= sphericity <= 1.`

File: src/Domain/CoordinateMaps/Frustum.hpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.hpp:189]: `\vec{\sigma}_{\mathrm{+z}} = ...`

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp]: Please put these all into the `Unit.Domain.CoordinateMaps.Frustum` test case.

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp:61]: CI fails here

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (nilsvu) [inline src/Domain/Creators/BinaryCompactObject.cpp:421]: I accidentally pushed a commit to this branch where I tried something, sorry for that! Not sure how it's even possible that I can push to a branch on your fork. Please feel free to just delete the commit.

I suggest here to elide the "first outer shell" entirely when it isn't needed anymore, i.e. when `frustum_sphericity == 1.` _and_ the user hasn't explicitly specified an `OuterShell.InnerRadius`. The commit I accidentally pushed does that, so feel free to take code from it if you want. I or you can also do this in a follow-up prefer if you prefer.

**Review by nilsvu — CHANGES_REQUESTED** [review]
I can confirm this works! With this map we can transition from the two inner cubes to a sphere in a single layer of blocks:

Bildschirmfoto 2021-10-18 um 00 11 44

File: src/Domain/Creators/BinaryCompactObject.cpp
Comment (kidder) [reply src/Domain/Creators/BinaryCompactObject.cpp:421]: Unless the creator of a PR unchecks a box, maintainers can push to a PR.

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: Is it true that the origin cannot be in a frustum? If I understand correctly, the inner cube that the frustums envelop can be offset from the origin far enough so the origin is in the frustum, e.g. in a high-mass-ration BBH domain?

Codecov actually complains that this case isn't covered by a test, so here it could be useful to add one?

File: src/Domain/CoordinateMaps/Frustum.hpp
Comment (nilsvu) [reply src/Domain/CoordinateMaps/Frustum.hpp:189]: ping on this

File: tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp]: CI wants you to add this option to a few input files in tests/ :)

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by nilsvu** [comment]
Please feel free to squash, including the comments above

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (mar-celine) [reply src/Domain/CoordinateMaps/Frustum.cpp]: With the latest tests I've done, it looks like the frustum can contain the origin, what was breaking the inverse tests was having the bulged out face be smaller than the opposite flat face. I will remove this line

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: `map(source_coords)`

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (nilsvu) [inline src/Domain/Creators/BinaryCompactObject.hpp]: Could you add a bit of info here? Help strings that just repeat the option name are not super helpful :D
You can also add `lower_bound` and `upper_bound` functions.

**Review by nilsvu — CHANGES_REQUESTED** [review]
Looks good, you can squash these in when fixing CI

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (knelli2) [inline src/Domain/CoordinateMaps/Frustum.cpp]: Could you add in what the current value of the sphericity is in the message? I find it helpful when debugging to know what the wrong value actually is without having to go look it up.

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (knelli2) [inline src/Domain/CoordinateMaps/Frustum.cpp]: `0.0`

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (knelli2) [inline src/Domain/CoordinateMaps/Frustum.cpp]: Could you add a test for this ASSERT as well?

**Review by knelli2 — CHANGES_REQUESTED** [review]
Just a couple things.

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: I suggest to keep the `if (with_equiangular_map)` instead of the new `make_with_value` (it introduces an unnecessary allocation and pointwise multiplication by one). This occurs a few more times below, also with `with_projective_map_`.

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml
Comment (nilsvu) [inline tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml]: Please keep this at `0.0` for now, just to make sure people using this input file don't get confused that the domain changed.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: Please change this one too

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: also here

File: src/Domain/CoordinateMaps/Frustum.cpp
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp]: and here

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by knelli2** [comment]
LGTM. You can squash after @nilsleiffischer gives his approval.

**Review by nilsvu — DISMISSED** [review]

**Comment by nilsvu** [comment]
yes please 👍

**Review by nilsvu — APPROVED** [review]

**Review by knelli2 — APPROVED** [review]

**Pull request merged**
